### PR TITLE
refactor: logging API

### PR DESCRIPTION
**Pull request by miurahr**
Related #715 

## Pull request type

- Other (describe below)

Refactor logging API and prepare for improvement of logging backend

## Which ticket is resolved?

- Title ex. fix...
  * URL ex. https://...
 
- Title
  * URL

Dev ML
https://sourceforge.net/p/omegat/mailman/message/37876949/

## What does this PR change?

- Introduce `o.o.u.logging.Logger`
- fix out of index error when custom logging configuration given
- make `o.o.u.Log` as a thin proxy of `o.o.u.l.Logger` instantiated with name "global"
- unit test
- fix usage of `Log.getLogFileLocation`  that result is not used
- rewrite `Log.logDebug` with new `Logger#logDebug` that remove dependency for JDK1.4 logging

### CODE REVIEW

**Comment by miurahr** [comment]
1. commit 67c1693d3a45b890bec7b852f7708b8173c805b2 introduce `o.o.u.logging.Logger` class that abstract JUL(java.util.Logger). 
2. commit d708dba update `o.o.u.Log` class to be a "global" logger using `o.o.u.l.Logger` and avoid code duplication.
3.  commit 8da3123  update all `Log.logDebug` to rewrite with `o.o.u.l.Logger` to avoid direct dependency for JUL.

**Comment by miurahr** [comment]
I've tried  to port `o.o.u.l.Logger` to use SLF4J instead of JUL.

https://github.com/omegat-org/omegat/pull/717/commits/10fd77e2072a36627a82e096ef75ec5764597156

File: src/org/omegat/util/logging/Logger.java [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+        if (!loaded) {
+            // load built-in logger settings
+            try (InputStream in = Log.class.getResourceAsStream("/org/omegat/logger.properties")) {
+                init(in);
+            } catch (IOException ex) {
+                logger.log(Level.SEVERE, "Can't open file for logging", ex);
+            }
+        }
+    }
+
```
Comment (miurahr) [inline src/org/omegat/util/logging/Logger.java:72]: We should load "logger.properties" only once.

File: src/org/omegat/util/logging/Logger.java [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+     */
+    public void log(String key, Object... parameters) {
+        if (logger.isLoggable(Level.INFO)) {
+            LogRecord rec = new LogRecord(Level.INFO, key);
+            rec.setResourceBundle(OStrings.getResourceBundle());
+            rec.setParameters(parameters);
+            rec.setLoggerName(logger.getName());
+            logger.log(rec);
+        }
+    }
```
Comment (miurahr) [inline src/org/omegat/util/logging/Logger.java:253]:
```suggestion
```

Unnecessary method.

**Review by t-cordonnier — COMMENTED** [review]
For the moment, I don't see what the class org.omegat.util.logging.Logger gives better than org.omegat.util.Log
You did a try to port it to SLF4J but finally it is not in the last version of your branch
Was it impossible to port org.omegat.util.Log to SLF4J?
(eventually forgetting the logDebug method as a first step, and considering it only after)

**Comment by miurahr** [comment]
> (that said, not sure it will require less changes to migrate from o.o.u.Log to o.o.u.l.Logger than directly to slf4j)

There are two custom requirement and one compatibility issue to port logging abstraction layer.

1. an API to get a log file path to provide  Help>Log... dialog
2. provide logDebug that give class/thread name in log file.
3. JUL template use "{0}" form, but SLF4J is "{ }". there need to template function.
 
> Now what happened with commit [10fd77e](https://github.com/omegat-org/omegat/commit/10fd77e2072a36627a82e096ef75ec5764597156)?

o.o.u.l.Logger is a thin wrapper of org.slf4j.Logger,  and give convenient methods
 such as `logErrorRB` and `getLogFileName`
It also has a private utility method to support "{0}" style from Bundle `getMessage`

**Comment by miurahr** [comment]
> (that said, not sure it will require less changes to migrate from o.o.u.Log to o.o.u.l.Logger than directly to slf4j)

I've tried at first to migrate directly to slf4j. It requires overhaul of all `Log.*` calls everywhere.
So I take a strategy to use a proxy design pattern to hide under logging layer from code space and replace it with slf4j.
If we can make  `o.o.u.Log` with slf4j directly, it will be better.

**Comment by miurahr** [comment]
I've tried to migrate directly SLF4J without making intermediate o.o.u.l.Logger class.
You can see.
https://github.com/omegat-org/omegat/pull/717

**Pull request closed without merging**